sound/soc/intel/boards/sof_rt5682.c

```diff
@@ -566,8 +568,36 @@ static struct snd_soc_dai_link *sof_card_dai_links_create(struct device *dev,
 			if (!links[id].cpus->dai_name)
 				goto devm_err;
 		}
+		id++;
 	}
 
+	links[id].name = "Compress Probe Playback";
+	links[id].id = id;
+	links[id].cpus = &cpus[id];
+	links[id].num_cpus = 1;
+	links[id].cpus->dai_name = "Probe Injection0 CPU DAI";
+	links[id].codecs = dummy;
+	links[id].num_codecs = 1;
+	links[id].platforms = platform_component;
+	links[id].num_platforms = ARRAY_SIZE(platform_component);
+	links[id].init = NULL;
+	links[id].ignore_suspend = 1;
+	links[id].nonatomic = 1;
+	id++;
+
+	links[id].name = "Compress Probe Capture";
+	links[id].id = id;
+	links[id].cpus = &cpus[id];
+	links[id].num_cpus = 1;
+	links[id].cpus->dai_name = "Probe Extraction CPU DAI";
+	links[id].codecs = dummy;
+	links[id].num_codecs = 1;
+	links[id].platforms = platform_component;
+	links[id].num_platforms = ARRAY_SIZE(platform_component);
+	links[id].init = NULL;
+	links[id].ignore_suspend = 1;
+	links[id].nonatomic = 1;
+
```

Comment:
we could avoid touching the machine drivers why doing topology tricks. I really don't want to change all machine drivers just for probes, that doesn't seem right.

Reply:
I had the exact opposite idea for cAVS. Removed the probes from topology entirely.
Instead, with ASoC multicodec:multiplatform support, one could be using PCI device (e.g.: HDA audio) as a bus with separate coards onboard:
- card0, let's say SSP card for rt274
- card1 for HDMI-only
- card2 probes

Looks much cleaner.
If you want to follow topology path, you have to modify all topologies instead.

Reply:
maybe the probes should be handled with MFD then.

Reply:
What MFD has ti do with this? Let's do the honest work and help ASoC folks with multicodec/ multiplatform support. MFD != multi machine board.

Back in Windows team, we were testing inter-op setups e.g: sdw + i2s + dmic + hda on CNL RVPs all enabled at once. You could also involve douple SSP codecs if required. Want Linux side to have similar capabilities.

Reply:
the MFD enables you to precisely do what you listed above

- card0, let's say SSP card for rt274
- card1 for HDMI-only
- card2 for SoundWire only
- card3 for DMIC
- card4 for probes

Also please watch your language and refrain from derogatory comments. You have agreed to comply with the Intel code of conduct and the kernel developer one as well. Final notice.

Reply:
Maybe I'm missing something, although what I saw is: register ssp_client. What happens if there are more than codecs ( > 1)? Same for sdw as there could be acatually MANY codecs there as well as sdw_aggr codecs. In general, my vision was to leave MFD for mutli-function stuff (from user perspective, e.g.: audio, video etc.) and then take advantage of multicodec::mutliplatform support from ASoC (provided it's there) for machine boards.

Reply:
> What happens if there are more ssp codecs ( > 1)?

All SSP codecs are handled by the same client. So the idea behind the MFD implementation is to really split into separate cards based on the DAI type. So all SSP's will be managed by the same card and the same machine driver declares the dai links for all the codecs. This makes it really simple to not have a monolithic topology file for all the PCM devices. Also making is easier to test for example a topology with just the HDMI pipelines, or just SDW etc

Reply:
And yes, this may not be the true definition of a traditional MFD but it gives us the same benefits without actually having to move the existing SOF code to drivers/mfd

Reply:
@ranj063 Ain't the "future" about having multiple topology files? I.e.: rt274 5 pipes, hdmi 3 pipes, rt286 2 pipes. During load, once machine boards are enumerated, sw-core assembles the WHOLE graph (with branches assigned to given boards).

When you launch given stream (aplay), FW instantiates module (and assigns an instance_id to it) and proceeds onward. This gives a very dynamic feeling to audio/mfd stack, overall.
Of course I'm using here many cAVS terms as I'm still young when it comes to SOF, but I'm sure you get the idea. If no, I can improve my description.

Reply:
> @ranj063 Ain't the "future" about having multiple topology files? I.e.: rt274 5 pipes, hdmi 3 pipes, rt286 2 pipes. During load, once machine boards are enumerated, sw-core assembles the WHOLE graph (with branches assigned to given boards).
Yes, absolutely. And this is where we are headed with the MFD. But I didnt quite get what you meant by "sw-core assembles the WHOLE graph". Yes, the FW is aware of all the entire graph ie the components and connections in all the topologies. But each client is isolated in terms of the graph it sets up.
> 
> When you launch given stream (aplay), FW instantiates module (and assigns an instance_id to it) and proceeds onward. This gives a very dynamic feeling to audio/mfd stack, overall.
> Of course I'm using here many cAVS terms as I'm still young when it comes to SOF, but I'm sure you get the idea. If no, I can improve my description.
This part I'm not sure about. The core does all the HW initialization prior to any of the clients being registered and needing to use it. Are you thinking about something like this for example: the i915 is init only if there HDMI client is being registered?

Reply:
> When you launch given stream (aplay), FW instantiates module (and assigns an instance_id to it) and proceeds onward. This gives a very dynamic feeling to audio/mfd stack, overall.

This part I'm not sure about. The core does all the HW initialization prior to any of the clients being registered and needing to use it. Are you thinking about something like this for example: the i915 is init only if there HDMI client is being registered?

Reply:
we did want to have one topology file per client indeed.
I am not sure however about 'assembling' the whole graph, this adds dependencies between clients that is not really needed. what matters is that the graph is consistent in each file.

Reply:
> sw-core assembles the WHOLE graph
So, software gets fed with multiple topology files. Rather than having some static declarations, everything would be dynamic (based on lists etc.). Each pipeline within such list would have an affinity field (owner, e.g.: rt274 provided description for this pipeline and is available for that board for use). Since you could have multiple ssp codecs (an example), you could have two topology files (one for rt274 and one for rt286). All the pipelines from all the topologies are combined by core part into one graph with so-called branches for each machine board.

"Assembling" this is again maybe not as clear because I'm coming from cAVS. You cannot have, let's say 100x copiers. FW limits you with certain number for each module.
Thanks @plbossart and @ranj063 for the quick responses. I'll need to spend more time on this, before moving forward with this.